lnwallet/chanfunding/coin_select.go

```diff
@@ -262,70 +252,85 @@ func CoinSelectSubtractFees(feeRate chainfee.SatPerKWeight, amt,
 	return selectedUtxos, outputAmt, changeAmt, nil
 }
 
-// CoinSelectUpToAmount attempts to select coins such that we'll spend up
-// to amt in total before fees, or all of the coins if not sufficient amount
-// available.
-func CoinSelectUpToAmount(feeRate chainfee.SatPerKWeight, maxAmt, reserved,
+// CoinSelectUpToAmount attempts to select coins such that we'll select up to
+// maxAmount exclusive fees and reserve if sufficient funds are available.
+// If insufficient funds are available this method selects all available coins.
+
+func CoinSelectUpToAmount(feeRate chainfee.SatPerKWeight, maxAmount, reserved,
 	dustLimit btcutil.Amount, coins []Coin) ([]Coin, btcutil.Amount,
 	btcutil.Amount, error) {
 
 	// Get balance from coins.
-	var balance btcutil.Amount
+	var totalBalance btcutil.Amount
 	for _, coin := range coins {
-		balance += btcutil.Amount(coin.Value)
+		totalBalance += btcutil.Amount(coin.Value)
 	}
 
-	// First we try to select coins up to the specified maxAmt.
-	outputAmt := maxAmt
+	// First we try to select coins to create an output of the specified
+	// maxAmount and a change output that covers the miner fee.
 	selectedCoins, changeAmt, err := CoinSelect(
-		feeRate, maxAmt, dustLimit, coins,
+		feeRate, maxAmount, dustLimit, coins,
 	)
-	if err == nil {
+
+	outputAmount := maxAmount
+
+	switch err.(type) {
+	// If that selection succeeds we check if our total balance covers
+	// the selected set of coins including fees plus an optional anchor
+	// reserve.
+	case nil:
 		// Validate that the selected amount will leave the required reserved
-		// value in the balance.
+		// value in our total balance.
+		// First we sum up the value of all selected coins.
 		selected := btcutil.Amount(0)
 		for _, coin := range selectedCoins {
 			selected += btcutil.Amount(coin.Value)
 		}
 
-		// We subtract the change amount to obtain the actual amount that is
-		// selected.
+		// We then subtract the change amount from the value of all selected
+		// coins to obtain the actual amount that is selected.
 		selected -= changeAmt
 
-		if balance-selected < reserved {
-			err = &errReservedValueInvalidated{reserved}
-		}
-	}
-	if err != nil {
-		switch err.(type) {
-		// Without sufficient funds we spend it all by subtracting
-		// the fee directly from the output amount.
-		case *ErrInsufficientFunds, *errReservedValueInvalidated:
-			selectedCoins, outputAmt, changeAmt, err = CoinSelectSubtractFees(
-				feeRate, balance-reserved, dustLimit, coins,
-			)
+		// Next we check if our total balance can cover for the selected
+		// output plus the optional anchor reserve.
+		if totalBalance-selected < reserved {
+			// If our local balance is insuffcient to cover for the reserve
+			// we try in the next step to select an output amount that uses
+			// our total balance minus reserve and fees.
+			selectedCoins, outputAmount, changeAmt, err =
```

Comment:
How about we have a boolean variable above the `switch` called `selectSubtractFees` that we set to true here and in case `case *ErrInsufficientFunds`.
Then after the switch block, we have an `if selectSubtractFees {` that calls this?